### DISCUSSION TITLE
How to copy / generate file to control-plane nodes before k3s server starts?

### DISCUSSION THREAD

**Discussion by jsimomaa** [mysticaltech/terraform-hcloud-kube-hetzner#665] [Q&A]
I read https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/issues/544 first but couldn't quite understand how one is able to copy a file to the control-plane nodes before k3s server is started?

My use case is to have an audit policy file and it is given to the api-server via:

```
--audit-policy-file=/path/to/audit/policy.yaml
```

https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/#audit-policy

I tried to look for existing issues or discussions related to this topic but could only find #544 . So please if someone could elaborate bit more how to actually achieve this, would be really appreciated 😄 

Thank you!

**Comment by mysticaltech** [comment]
@jsimomaa Look how it's done for instance here https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/9bc31b9fc081cd20805b7a803b7b136cad11c852/control_planes.tf#L94

You have to use the File provisioner https://developer.hashicorp.com/terraform/language/resources/provisioners/file.

That would require you to clone the repo and edit it, or if you prefer to not clone the repo and use the module as is, you can use the `preinstall_exec` variable to fetch that file from a specific location.

Then you can use the `control_planes_custom_config` to pass the audit-policy-file option above.

**Reply by jsimomaa** [reply]
Thank you for the fast reply!

I was unsure should I clone the repo and edit because I would like to use the module as is and do the file copying without editing the module. But your answer clears up the confusion in my head

I will check how `preinstall_exec` shouold be applied here
